### PR TITLE
Optimize param validation through `get_param(...) -> Option<Out>`

### LINKED ISSUE #15505: Optimize `validate_param`
## What problem does this solve or what need does it fill?

From discord discussion: https://discord.com/channels/691052431525675048/749335865876021248/1289681891611508809

As @cart pointed out, `validate_param` doubles lookup, which can have significant performance impact.

## What solution would you like?

There are few potential ways of fixing that:
1. Merge `get_param` with `validate_param` into `try_get_param`.
1.1. In naive approach, we always create a system task (multithreaded executor), even if parameters can't be retrieved.
1.2. We can try to run `try_get_param` before spawning a task, but then we need to somehow pass the parameters into the system. (type erasure issues)
2. Keep `get_param` and `validate_param`, add `try_get_param` with default implementation that merges the previous 2.
Implement manually for performance gains.

## Additional context

Blocked by #15394, we want most machinery in place before we make this change.

### PR DESCRIPTION
**Pull request by MiniaczQ**
# Objective

Fixes #15505 currently through 1.1, experimenting with 1.2.
Alternative to #15571

## Solution

- Describe the solution used to achieve the objective above.

## Testing

- Did you test these changes? If so, how?
- Are there any parts that need more testing?
- How can other people (reviewers) test your changes? Is there anything specific they need to know?
- If relevant, what platforms did you test these changes on, and are there any important ones you can't test?

---

## Showcase

> This section is optional. If this PR does not include a visual change or does not add a new feature, you can delete this section.

- Help others understand the result of this PR by showcasing your awesome work!
- If this PR adds a new feature or public API, consider adding a brief pseudo-code snippet of it in action
- If this PR includes a visual change, consider adding a screenshot, GIF, or video
  - If you want, you could even include a before/after comparison!
- If the Migration Guide adequately covers the changes, you can delete this section

While a showcase should aim to be brief and digestible, you can use a toggleable section to save space on longer showcases:

<details>
  <summary>Click to view showcase</summary>

```rust
println!("My super cool code.");
```

</details>

## Migration Guide

> This section is optional. If there are no breaking changes, you can delete this section.

- If this PR is a breaking change (relative to the last release of Bevy), describe how a user might need to migrate their code to support these changes
- Simply adding new functionality is not a breaking change.
- Fixing behavior that was definitely a bug, rather than a questionable design choice is not a breaking change.

### CODE REVIEW

**Comment by MiniaczQ** [comment]
The steps taken to get here look as following:
1. We question whether 1.2 (from #15505) is possible to implement. While lifetimes and type erasure can be worked around, we can't fetch all system parameters, because in case of compound systems (e.g. `a.pipe(b)`), they can have exclusive parameters. While this isn't technically UB, since the parameters will only be used in mutually exclusive contexts, this will possibly lead to multiple mutable/immutable references to the same data existing at once.
2. To get to 1.2, we need proposition 1.1 anyways. It gives better internal ergonomics than current solution. Because it's only applied when running systems, we don't save on tasks in multithreaded executors nor do we retain the original behavior of running only entire compound systems in executors (e.g. `a.pipe(b)` could run `a` but stop at `b` since params were "checked" lazily).
3. We can fix that by reintroducing `validate_param`, but default it's implementation as `get_param().is_some()`. We expect this implementation to be optimized by compiler. Because default implementation breaks for non-send resources, we overwrite it as well as compound resources (`SystemParam` tuples, `ParamSet`, `DynSystemParam`) to rely on their sub-params implementations.

To sum it up, without going the additional step into proposition 1.2 we simply cannot reduce double checks in executors.
The updated idea can take 2 possible forms:
a) `System`s and `SystemParam`s `get_param()` will return optional resource and multiple conflicting references will exist (this may tip miri off)
b) We separate it into `validate_param() -> Option<Accessor>` which can validate the access without constructing the references and `get_param(Accessor)` which will construct the reference.

**Comment by MiniaczQ** [comment]
Marking it as ready for review since code is basically done, just needs benchmarks

**Comment by MiniaczQ** [comment]
Note for @alice-i-cecile, since I've only made groundwork towards 1.2, but not implemented it, this doesn't have a significant impact on performance and we can delay it.
I've had trouble finding all the broken code, but now that I've done it once it should go much quicker

**Comment by MiniaczQ** [comment]
I found an edge case that asks for change in requirements:
https://discord.com/channels/691052431525675048/749335865876021248/1292140022375514145
I'm re-drafting this until I implement it

**Comment by MiniaczQ** [comment]
Core changes are done, just need to update documentation

**Comment by MiniaczQ** [comment]
Before (`main`):
```
param/combinator_system/8_piped_systems
                        time:   [1.6549 µs 1.6689 µs 1.6852 µs]
                        change: [-13.250% -10.268% -7.3630%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

param/combinator_system/8_dyn_params_system
                        time:   [95.545 ns 95.841 ns 96.172 ns]
                        change: [-2.9593% -1.3542% -0.0548%] (p = 0.06 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

param/combinator_system/8_variant_param_set_system
                        time:   [91.596 ns 91.962 ns 92.365 ns]
                        change: [+0.1790% +1.0297% +1.8957%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  6 (6.00%) high mild
  3 (3.00%) high severe
  ```
After:
```
param/combinator_system/8_piped_systems
                        time:   [1.6590 µs 1.6863 µs 1.7149 µs]
                        change: [-3.5415% -1.5778% +0.5378%] (p = 0.15 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild

param/combinator_system/8_dyn_params_system
                        time:   [93.629 ns 94.065 ns 94.551 ns]
                        change: [-2.3199% -1.2717% -0.1379%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

param/combinator_system/8_variant_param_set_system
                        time:   [91.913 ns 92.361 ns 92.842 ns]
                        change: [+0.0434% +0.8115% +1.6509%] (p = 0.05 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
```

There is a lot of noise in the measurements, but I at the very least I don't think this is a regression

**Comment by MiniaczQ** [comment]
> @MiniaczQ, this looks like it should be reviewable right? There's a _lot_ of merge conflicts now though 😅

It used to be, all the ideas are here, it just affects such basic APIs that everything conflicts.
I don't have time to resolve this right now, there is a good chance that starting from scratch will be faster, since most of the changes can be regexed

**Comment by chescock** [comment]
The changes look reasonable, but it seems like a rebase would change enough that I should wait before clicking Approve.  

I don't quite follow the motivation.  It seems like the original goal was to avoid a double lookup, but then you determined that wouldn't be possible.  Is the current goal to remove duplication between `Single::get_param` and `Single::validate_param` and consolidate the calls to `try_warn_param` in one place?  Oh, and to change the behavior of `PipeSystem` so that the second system's parameters are only validated after the first system runs?  

The changes to `SystemState::get` seem unfortunate.  That's a lot of `unwrap()`s!  Is the reasoning that cases like `SystemState::<Single>::get` might now fail, and we need a way to express that?  

Do we expect anyone use fallible parameters with `SystemState`?  The advantage in regular systems is that it gives a way to declare a run condition as part of the system definition, but that doesn't apply to `SystemState`.  I'd vote for leaving `SystemState::get` as panicking so that uses who don't use fallible parameters don't need to `unwrap`, and then exposing a `try_get` for users who do.  (Or, if we really want to reduce accidental panics, we could add `trait InfallibleSystemParam: SystemParam` and bound `get` on it.)

**Comment by MiniaczQ** [comment]
> I don't quite follow the motivation. It seems like the original goal was to avoid a double lookup, but then you determined that wouldn't be possible. Is the current goal to remove duplication between `Single::get_param` and `Single::validate_param` and consolidate the calls to `try_warn_param` in one place? Oh, and to change the behavior of `PipeSystem` so that the second system's parameters are only validated after the first system runs?

The core issue is that we cannot validate params anywhere, but directly before running a system. In case of `a.pipe(b)` system `a` can invalidate `b`s params. With this restraint, double lookup is no longer an issue, since we don't check `b`s params until we run it.

> The changes to `SystemState::get` seem unfortunate. That's a lot of `unwrap()`s! Is the reasoning that cases like `SystemState::<Single>::get` might now fail, and we need a way to express that?

It always could have failed, but we agreed a failure results in a panic. There aren't many unwraps in the core code, since failures are forwarded to `System::run`.
It's mostly the test cases that suffer (fully justifiable) and very few inner API consumers, which unwrap under the contract that all the passes system arguments are available.

> Do we expect anyone use fallible parameters with `SystemState`? The advantage in regular systems is that it gives a way to declare a run condition as part of the system definition, but that doesn't apply to `SystemState`. I'd vote for leaving `SystemState::get` as panicking so that uses who don't use fallible parameters don't need to `unwrap`, and then exposing a `try_get` for users who do. (Or, if we really want to reduce accidental panics, we could add `trait InfallibleSystemParam: SystemParam` and bound `get` on it.)

The exact API wasn't agreed upon, i guess this is a good split point for the rewrite, since it drastically reduces the API changes. If we want non-panicking default we can do a followup.
As for `SystemState`, it's still an API that can fail and users should be able to handle that.

**Comment by chescock** [comment]
> The core issue is that we cannot validate params anywhere, but directly before running a system. In case of `a.pipe(b)` system `a` can invalidate `b`s params. With this restraint, double lookup is no longer an issue, since we don't check `b`s params until we run it.

Yup, that makes sense!  I think I'm just asking to update the "Objective" section in the PR.  It links to an issue that talks about getting rid of all of the double lookups, but this PR doesn't get rid of them for ordinary systems in the multi-threaded executor, so it's not immediately clear what the goal *is* here.  

> It always could have failed, but we agreed a failure results in a panic.

It depends on the parameter, though, right?  Like, `SystemState::<Single>::get` can fail, but `SystemState::<Query>::get` will *always* return `Some`.  If you know you have a `Query`, it would feel bad to have to add an extra `unwrap()`!  And my impression is that most uses of `SystemState` are with infallible parameters like that.  

The difference I see between this and things like `Query::single` and `World::resource` are that `SystemParam` fallibility is part of the *type* rather than an assertion about the runtime state of the program.  

In the future, we might even be able to add a `InfallibleSystemParam` subtrait that returns the param without an `Option`, similar to how `Ord` is a subtrait of `PartialOrd` that returns an `Ordering` without an `Option`.  We could have `SystemState::try_get` that works on any param and returns an `Option`, and `SystemState::get` that works only on infallible params and returns the actual param.  If we do want to go down that path, then forcing users to add an `unwrap()` now and then remove it later would be extra churn.  

I'm not suggesting that any of that should be in this PR, of course!  But I think it might be worth making this PR *smaller* by moving the `unwrap()` into `SystemParam::get` so we only need to write it once.  It's no worse than what we have now, and it doesn't require updating every caller.

File: crates/bevy_ecs/macros/src/lib.rs [hunk 1]
```diff
@@ -316,9 +316,11 @@ pub fn impl_param_set(_input: TokenStream) -> TokenStream {
                 // SAFETY: systems run without conflicts with other systems.
                 // Conflicting params in ParamSet are not accessible at the same time
                 // ParamSets are guaranteed to not conflict with other SystemParams
-                unsafe {
+                let maybe_param = unsafe {
                     #param::get_param(&mut self.param_states.#index, &self.system_meta, self.world, self.change_tick)
-                }
+                };
+                // `ParamSet::get_param` ensures all sub-params are accessible.
```
[the remark this answers is not in the document]
Comment (MiniaczQ) [reply crates/bevy_ecs/macros/src/lib.rs:322]: Can't tell now, this was made quite a while ago.
We'll definitely want someone experience check the ParamSet, DynamicSystemParam, etc.

**Comment by MiniaczQ** [comment]
> It depends on the parameter, though, right? Like, `SystemState::<Single>::get` can fail, but `SystemState::<Query>::get` will _always_ return `Some`. If you know you have a `Query`, it would feel bad to have to add an extra `unwrap()`! And my impression is that most uses of `SystemState` are with infallible parameters like that.

We try to store param's ability to fail in type and design an API around that.
Not sure if that's worth the effort and boilerplate though.

Since this PR I've read some discussion about fallibility and few new decisions include removing per-system param fail behavior.
This should be a app-global behavior or a wrapper around individual system params.
We can account for the latter case in w/e API this ends up being.

Also, since this is a pretty old PR with lots of outdated information and crap ton of merge conflicts, I'd rather have it closed whenever someone picks up the overall issue.
